core/state/state_witness.go

```diff
@@ -0,0 +1,216 @@
+		var ownerPaths []string
+		var ownerNodes [][]byte
+
+		for path, node := range nodeMap {
+			ownerPaths = append(ownerPaths, path)
+			ownerNodes = append(ownerNodes, node)
+		}
+		encWit.AllPaths = append(encWit.AllPaths, ownerPaths)
+		encWit.AllNodes = append(encWit.AllNodes, ownerNodes)
+	}
+
+	for codeHash, code := range w.codes {
+		encWit.CodeHashes = append(encWit.CodeHashes, codeHash)
+		encWit.Codes = append(encWit.Codes, code)
+	}
+
+	for blockNum, blockHash := range w.blockHashes {
+		encWit.BlockNums = append(encWit.BlockNums, blockNum)
+		encWit.BlockHashes = append(encWit.BlockHashes, blockHash)
+	}
+	res, err := rlp.EncodeToBytes(&encWit)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (w *Witness) addAccessList(owner common.Hash, list map[string][]byte) {
+	if len(list) > 0 {
+		w.lists[owner] = list
+	}
+}
+
+func (w *Witness) AddBlockHash(hash common.Hash, num uint64) {
+	w.blockHashes[num] = hash
+}
+
+// TODO: don't include the code hash in the witness if not necessary
+func (w *Witness) AddCode(hash common.Hash, code Code) {
+	if code, ok := w.codes[hash]; ok && len(code) > 0 {
+		return
+	}
+	w.codes[hash] = code
+}
+
+func (w *Witness) AddCodeHash(hash common.Hash) {
+	if _, ok := w.codes[hash]; ok {
+		return
+	}
+	w.codes[hash] = []byte{}
+}
+
+func (w *Witness) LogSizeWithBlock(b *types.Block) {
+	enc, _ := w.EncodeRLP()
+	fmt.Printf("block %d witness+block size: %d\n", b.Number(), len(enc))
+}
+
+func (w *Witness) Summary() string {
+	b := new(bytes.Buffer)
+	xx, err := rlp.EncodeToBytes(w.block)
+	if err != nil {
+		panic(err)
+	}
+	totBlock := len(xx)
+
+	yy, _ := w.EncodeRLP()
+
+	totWit := len(yy)
+	totCode := 0
+	for _, c := range w.codes {
+		totCode += len(c)
+	}
+	totNodes := 0
+	totPaths := 0
+	nodePathCount := 0
+	for _, ownerPaths := range w.lists {
+		for path, node := range ownerPaths {
+			nodePathCount++
+			totNodes += len(node)
+			totPaths += len(path)
+		}
+	}
+
+	fmt.Fprintf(b, "%4d hashes: %v\n", len(w.blockHashes), common.StorageSize(len(w.blockHashes)*32))
+	fmt.Fprintf(b, "%4d owners: %v\n", len(w.lists), common.StorageSize(len(w.lists)*32))
+	fmt.Fprintf(b, "%4d nodes:  %v\n", nodePathCount, common.StorageSize(totNodes))
+	fmt.Fprintf(b, "%4d paths:  %v\n", nodePathCount, common.StorageSize(totPaths))
+	fmt.Fprintf(b, "%4d codes:  %v\n", len(w.codes), common.StorageSize(totCode))
+	fmt.Fprintf(b, "%4d codeHashes: %v\n", len(w.codes), common.StorageSize(len(w.codes)*32))
+	fmt.Fprintf(b, "block (%4d txs): %v\n", len(w.block.Transactions()), common.StorageSize(totBlock))
+	fmt.Fprintf(b, "Total size: %v\n ", common.StorageSize(totWit))
+	return b.String()
+}
+
+func (w *Witness) SetBlock(b *types.Block) {
+	w.block = b
+}
+
+func (w *Witness) Copy() *Witness {
+	var res Witness
+	res.block = w.block // we don't actually mutate the block in the witness so don't deep copy
+
+	for blockNr, blockHash := range w.blockHashes {
+		res.blockHashes[blockNr] = blockHash
```

Comment:
I think this will panic, because the res.blockHashes is not initialized. e.g. `res.blockHashes = make(map[uint64]common.Hash, 0, len(w.blockHashes)`